docs/enhancements/dynamic-fetching-of-openshift-versions.md

```diff
@@ -0,0 +1,301 @@
+
+## Proposal
+
+The assisted-service will make use of a new configuration variable called
+``DYNAMIC_OPENSHIFT_RELEASES_CONFIG``, which outlines the supported matrix. For
+example:
+
+```
+DYNAMIC_OPENSHIFT_RELEASES_CONFIG='{
+   "4.11": {
+      "upgrade_channels": [
+         {
+            "cpu_architecture": "x86_64",
+            "channels": ["stable"]
+         }
+      ]
+   },
+   "4.12": {
+      "upgrade_channels": [
+         {
+            "cpu_architecture": "x86_64",
+            "channels": ["candidate", "stable"]
+         },
+         {
+            "cpu_architecture": "arm64",
+            "channels": ["candidate", "stable"]
+         },
+         {
+            "cpu_architecture": "multi",
+            "channels": ["candidate", "stable"]
+         }
+      ]
+   }
+}'
+```
+
+The assisted-service periodically polls versions from the ``upgrades_info`` API
+in the form of ``https://api.openshift.com/api/upgrades_info/v1/graph?channel=<channel>&arch=<arch>``.
+For example:
+```
+$ curl -s 'https://api.openshift.com/api/upgrades_info/v1/graph?channel=stable-4.11&arch=amd64' \
+    | jq -r '.nodes[] | select(.version | startswith("4.11")) | .version'
+4.11.9
+4.11.4
+4.11.8
+...
+```
+
+It then adds any new version found by this process in the database. When a
+consumer of the API requests the available versions via ``/v2/openshift-versions``
+(either it's the UI or a direct REST request) versions are read from the
+database.
+
+To help fetching only certain versions via auto-completion like U/X, we should
+provide a parameter ``version_pattern`` for the ``/v2/openshift-versions``
+endpoint, along with a ``only_latest`` parameter. For example:
+```
+/v2/openshift-versions:
+   get:
+      ...
+      parameters:
+        - in: query
+          name: version_pattern
+          description: Version pattern for the returned releases. E.g. '4.12'.
+          type: string
+          required: false
+        - in: query
+          name: only_latest
+          description: It true, returns only the latest version for each minor.
+          type: boolean
+          allowEmptyValue: true
+          default: false
+```
+
+In addition, we should expose a new API to show the current configuration
+of dynamic releases for debugging and transparency purposes.
+
+We will still need to handle each new minor version (x.y) by:
+* adding RHCOS images for the new version for each supported CPU architecture.
+* adding x.y to the new ``DYNAMIC_OPENSHIFT_RELEASES_CONFIG`` configuration.
+* setting the relevant feature-levels in feature-support-levels list.
+
+This way the assisted-service can support each x.y.z version by fetching a
+matching x.y RHCOS version from the assisted-image-service. The
+assisted-image-service stays up and ready to serve all supported RHCOS images
+and the amount of images is only the bare minimum.
+We assume here that each x.y RHCOS version is backwards and forwards compatible
+with all x.y.z OpenShift versions.
```

Comment:
I think it may be worth highlighting this somewhere more visible. Maybe in `Implementation Details/Notes/Constraints`?

Reply:
Done!